crates/sui-types/src/messages_checkpoint.rs

```diff
@@ -119,10 +124,149 @@ impl Default for ECMHLiveObjectSetDigest {
     }
 }
 
+// TODO: Does this need versioning?
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ObjectCheckpointState {
+    pub id: ObjectID,
+    /// This could be None if the object was deleted or wrapped in the transaction.
+    pub digest: Option<ObjectDigest>,
+}
+
+impl ObjectCheckpointState {
+    pub fn new(id: ObjectID, digest: Option<ObjectDigest>) -> Self {
+        Self { id, digest }
+    }
+}
+
+impl PartialEq for ObjectCheckpointState {
+    fn eq(&self, other: &Self) -> bool {
+        self.id == other.id
+    }
+}
+
+impl Eq for ObjectCheckpointState {}
+
+impl PartialOrd for ObjectCheckpointState {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for ObjectCheckpointState {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.id.cmp(&other.id)
+    }
+}
+
+/// A sorted list of object checkpoint states (by object ID).
+#[derive(Debug)]
+pub struct ObjectCheckpointStates {
+    pub contents: Vec<ObjectCheckpointState>,
+}
+
+impl ObjectCheckpointStates {
+    pub fn new(contents: BTreeMap<ObjectID, Option<ObjectDigest>>) -> Self {
+        Self {
+            contents: contents
+                .into_iter()
+                .map(|(id, digest)| ObjectCheckpointState::new(id, digest))
+                .collect::<Vec<_>>(),
+        }
+    }
+
+    /// Compute the Merkle root of the object checkpoint states.
+    /// Assumes that the object checkpoint states are sorted by object ID.
+    pub fn digest(&self) -> SuiResult<ObjectCheckpointStatesDigest> {
+        let tree = MerkleTree::<Blake2b256>::build_from_unserialized(self.contents.iter())
+            .map_err(|e| SuiError::GenericAuthorityError {
+                error: format!("Failed to build Merkle tree: {}", e),
+            })?;
+        let root = tree.root().bytes();
+        Ok(ObjectCheckpointStatesDigest::from(root))
+    }
+}
+
+#[derive(Debug)]
+pub struct ObjectCheckpointStatesDigest {
+    pub digest: Digest,
+}
+
+impl From<[u8; 32]> for ObjectCheckpointStatesDigest {
+    fn from(digest: [u8; 32]) -> Self {
+        Self {
+            digest: Digest::new(digest),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct CheckpointArtifacts {
+    pub latest_object_states: ObjectCheckpointStates,
+    // Add more.. e.g., execution digests, events, etc.
```

Comment:
While we can evolve this struct since it is not stored on chain, there is another problem: since individual struct fields don't have intent tags, how will we prevent extension attacks if we add other fields?

I think a safer approach is to have

```
pub enum CheckpointArtifacts {
    ObjectCheckpointStates(ObjectCheckpointStates),
    // add new variants here
}
```

and then the commitment is a hash of a vector of these